### ISSUE TITLE
State

### ISSUE THREAD

**Issue by mottosso** [pyblish/pyblish-qml#60]
# Goal

Facilitate changes being made to data in GUI.

At the moment, data is read from a host (e.g. Maya) and sent to QML. But it is never returned.

That's allright as we currently can't modify the data. But we want to be able to, which means we'll need a method of sending the modified data back to the host.

The same goes for toggling instances. A toggled instance either has no `publish` data member, or it exists but is set to `False`. 

Toggling an item must create this attribute and set it to `True`. And this data must also be sent back to the host.

### State

The `state` consist of *everything* visualised by a GUI, including data that may be modified.

When the time comes to publish, the entirety of the `state` is sent back and used to re-create the state from the GUI in the host prior to publishing; like which items are toggled, any modified instance data etc.

```
 _________                _________
|         |   initial    |         |
|  State  |-------------->         |
|_________|              |         |
                         |   Gui   |
 _________               |         |
|         |   modified   |         |
|  State  <--------------|         |
|_________|              |_________|
```

# Implementation

Distil all that passed over-the-wire into a single data-structure; a JSON-serialisable dictionary. The dictionary is then (optionally) altered from within the GUI and sent back to be de-serialised.

**Example data structure**

```json
{
  "context": {
       "data": 
           {
               "key": "value"
           },
       "children": [
           {
               "name": "MyInstanceA",
               "objName": "plugin_SEL",
               "children": [],
               "data": {}
           }
       ]
  },
  "plugins": [
    {
       "name": "PluginA",
       "version": [0, 1, 0],
       "family": "myFamily",
       "category": "Mesh",
       "requires": ">=1.0.2",
       "order": 1,
       "absoluteOrder": 0,
       "optional": true,
       "doc": "Description",
     },
  ]
}
```

**Comment by mottosso** [comment]
### Changes

Ok, one caveat of this is; how do we deal with data that cannot be sent to the GUI, when that data is then sent back?

For example; is an instance contains data of type `PyNode`. We can't serialise it, so we will either have to skip it or pass a representation of is - such as a string. Once the data is then sent back to the host, that particular data will then either not be included or be incomplete.

The idea was for the data to be passed back in the exact shape and form as would be validated; including all it's members and data. That way, the user would have the option of modifying this data prior to sending it back, thus altering the publish.

What we'll have to do instead is to send what we can, and return *changes* to that data.

The user will then only be able to modify data that passed through; and the host will only be updating data that was changed, thus keeping track of non-serialisable data like `PyNode`s.

### Tracking changes

But, how does one keep track of changes?

The data is currently stored within a `QAbstractListModel`; data is stored directly into the members of a instance of `Item` which inherits `object`.

```python
class Item(object):
    def __init__(self):
        self.name = None
        self.family = None
        ...
```

They are each populated with the given state.

```python
item = Item()
item.name = state["name"]
item.family = state["family"]
```

But, in order for us to track which have changed and which have not, we would need additional data along with each value. We could, for example, wrap each value into an object to keep track of when it changes.

```python
class Value(object):
    def __init__(self, value):
        self.value = value
        self.changed = False

    def set(self, value):
        self.value = value
        self.changed = True

item = Item()
item.name = Value(state["name"])
item.family = Value(state["family"])
```

Now whenever a value is changed, it's `changed` flag will be flipped, allowing us to query and return only values that have changed.

```python
changed = list(item for item in model if item.changed)
```

### Transaction overview

Thus, the data going in and out may look something like this.

**Input**

```json
{
    "next_instance": null,
    "next_plugin": null,
    "current_plugin": null,
    "current_instance": null,
    "context": {
        "data": {
            "pyblishVersion": "1.0.13",
            "connectTime": 1426109222.349,
            "pythonVersion": "2.7.9 (default, Dec 10 2014, 12:28:03) [MSC v.1500 64 bit (AMD64)]",
            "host": "python",
            "user": "marcus",
            "endpointVersion": "1.1.4",
            "port": -1
        },
        "children": [
            {
                "name": "Peter01",
                "family": "napoleon.asset.rig",
                "publish": false,
                "objName": "Peter01",
                "data": {
                    "family": "napoleon.asset.rig",
                    "destination": "/server/published/assets",
                    "assetSource": "/server/assets/Peter",
                    "publish": false,
                    "minWidth": 800,
                    "identifier": "napoleon.instance"
                },
                "children": [
                    "node1",
                    "node2",
                    "node3"
                ]
            },
            {
                "name": "Richard05",
                "family": "napoleon.animation.cache",
                "publish": true,
                "objName": "Richard05",
                "data": {
                    "family": "napoleon.animation.cache",
                    "destination": "/server/published/assets",
                    "assetSource": "/server/assets/Peter",
                    "publish": true,
                    "minWidth": 800,
                    "identifier": "napoleon.instance"
                },
                "children": [
                    "node1",
                    "node2",
                    "node3"
                ]
            },
            {
                "name": "Steven11",
                "family": "napoleon.animation.cache",
                "publish": true,
                "objName": "Steven11",
                "data": {
                    "family": "napoleon.animation.cache",
                    "destination": "/server/published/assets",
                    "assetSource": "/server/assets/Peter",
                    "publish": true,
                    "minWidth": 800,
                    "identifier": "napoleon.instance"
                },
                "children": [
                    "node1",
                    "node2",
                    "node3"
                ]
            },
            {
                "name": "Piraya12",
                "family": "napoleon.animation.cache",
                "publish": true,
                "objName": "Piraya12",
                "data": {
                    "family": "napoleon.animation.cache",
                    "destination": "/server/published/assets",
                    "assetSource": "/server/assets/Peter",
                    "publish": true,
                    "minWidth": 800,
                    "identifier": "napoleon.instance"
                },
                "children": [
                    "node1",
                    "node2",
                    "node3"
                ]
            },
            {
                "name": "Marcus",
                "family": "napoleon.animation.cache",
                "publish": true,
                "objName": "Marcus",
                "data": {
                    "family": "napoleon.animation.cache",
                    "destination": "/server/published/assets",
                    "assetSource": "/server/assets/Peter",
                    "publish": true,
                    "minWidth": 800,
                    "identifier": "napoleon.instance"
                },
                "children": [
                    "node1",
                    "node2",
                    "node3"
                ]
            }
        ]
    },
    "plugins": [
        {
            "category": null,
            "families": [
                "napoleon.animation.cache"
            ],
            "name": "ExtractAsMa",
            "hasRepair": false,
            "optional": false,
            "type": "Extractor",
            "version": [
                0,
                0,
                0
            ],
            "hosts": [
                "python"
            ],
            "doc": null,
            "active": true,
            "requires": "pyblish>=1",
            "order": 2
        },
        {
            "category": null,
            "families": [
                "napoleon.animation.cache"
            ],
            "name": "ConformAsset",
            "hasRepair": false,
            "optional": false,
            "type": "Conformer",
            "version": [
                0,
                0,
                0
            ],
            "hosts": [
                "python"
            ],
            "doc": null,
            "active": true,
            "requires": "pyblish>=1",
            "order": 3
        },
        {
            "category": null,
            "families": [
                "napoleon.animation.cache"
            ],
            "name": "ValidateFailureMock",
            "hasRepair": false,
            "optional": true,
            "type": "Validator",
            "version": [
                0,
                0,
                1
            ],
            "hosts": [
                "python"
            ],
            "doc": null,
            "active": true,
            "requires": "pyblish>=1",
            "order": 1
        },
        {
            "category": null,
            "families": [
                "napoleon.animation.cache"
            ],
            "name": "ValidateNamespace",
            "hasRepair": false,
            "optional": false,
            "type": "Validator",
            "version": [
                0,
                0,
                1
            ],
            "hosts": [
                "python"
            ],
            "doc": null,
            "active": true,
            "requires": "pyblish>=1",
            "order": 1
        },
        {
            "category": null,
            "families": [
                "napoleon.animation.cache"
            ],
            "name": "ValidateFailureMock",
            "hasRepair": false,
            "optional": true,
            "type": "Validator",
            "version": [
                0,
                0,
                1
            ],
            "hosts": [
                "python"
            ],
            "doc": null,
            "active": true,
            "requires": "pyblish>=1",
            "order": 1
        },
        {
            "category": null,
            "families": [
                "napoleon.animation.cache"
            ],
            "name": "ValidateNamespace",
            "hasRepair": false,
            "optional": false,
            "type": "Validator",
            "version": [
                0,
                0,
                1
            ],
            "hosts": [
                "python"
            ],
            "doc": null,
            "active": true,
            "requires": "pyblish>=1",
            "order": 1
        }
    ]
}
```

**Output**

```json
{
    "context": {
        "children": [
            {
                "name": "Peter01",
                "data": {
                    "publish": false
                }
            }
        ]
    }
}
```

Indicating that the only thing that changed in the GUI was that the `Instance` called `Peter01` was toggled off and should *not* be published.

**Comment by mottosso** [comment]
### New Data

The above may work well for when we're only looking for modify existing data. It may also work well for removing data. But what about *adding* it?

For example, adding a new instance or plug-in from the GUI; with this approach we could send the resulting data of this new instance/plug-in back to the host, but the host would be unable to simply create an instance or plug-in with that data alone. A `Plugin` consists of an actual Python object that does things, whereas an `Instance` is merely a representation of physical data within the host.

But hold on - what is the use-case for creating a new plug-in or instance from the GUI? 

We know that:

1. Plug-ins are written as Python classes and loaded into Pyblish during discovery.
2. Instances are created within the Host and *visualised* by the GUI

(1) suggests that to fully create a plug-in from the GUI, one would have to provide for a text-editor and file-saving capabilities, whereas (2) suggests that we would need to use Pyblish to *create new* data, and not merely validate and extract it. However that goes outside of its responsibilities.

### Conclusion

In effect, all that the GUI could possibly do is to *change* existing data, not create it. This simplifies our implementation quite a bit.

**Comment by mottosso** [comment]
An update on this; state is passed and updated as expected.

To reach feature paridy, the only remaining factor is re-adapting the GUI to make use of how the data is formatted differently. I've also included a hint for which plug-in and instance will be processed next, allowing us to properly display the currently running plug-in*.

*Previously, we could only send a request to say "advance processing" which would look at the list of plug-ins and process the next one in line. Due to this, there was no way for us to know for certain which plug-in was being processed until after processing had completed; thus only being able to show *what has been done* and not *what is happening* right now or will happen next.

**Comment by mottosso** [comment]
### Tracking changes 2

An alternative to keeping track of each change within the attribute itself is to jot it down into a separate dictionary.

1. Whenever a value is changed
2. Store key with current and previous value.

**changes.json**

```json
[
  {
    "item": "Peter01",
    "changes": {
       "destination": {"old": "/server/published/assets", "new": "c:/assets"},
       "startFrame": {"old": 1001, "new": 1203},
       "publish": {"old": false, "new": true}
    }
  }
]
```

### Storage

The changes can then be stored either on a per-model basis, or per-application. The immediate benefit of storing it per-model is the direct access between a model and it's changes as we would otherwise have to emit changes that get picked up by the controller. On the other hand, the controller is the one making use of these changes and so it might be more efficient, and make more sense, to store it there to begin with.

The controller can then also maintain history so as to facilitate undo/redo, which is something we would otherwise have to synchronise if stored separately within each model.

Another advantage to this approach is the immediate access to the full list of changes made. The former approach would require a serialisation process to take place which may or may not be more efficient, depending on how often we choose to make use of the data. Time will tell.

**Comment by mottosso** [comment]
### Reliable Representation

A deadly caveat of this approach is the dependency on that the representation must be reliable.

For example, if a plug-in is misrepresented or lost, no changes can be recorded for it and thus it's state can not be changed. If it's state then defaults to "Run", then:

1. We will not expect it to run, as it it not visible
2. It will run, producing results we won't understand

In a more obscure area; imagine a value - `startFrame` - of type `MFloat`. A value existing in Maya and looks like any other number. However in the GUI the value cannot be represented. It defaults to 1000, but we've set our timeline to 900. Now whenever we publish, we would expect it to publish from frame 900 and we won't get any indication of it doing anything else. It will however publish from frame 1000, leaving us to go bug hunting.

**Comment by mottosso** [comment]
### Current Advance Model

..and why it might not work.

This is how processes are currently triggered from GUI to host.

![image](https://cloud.githubusercontent.com/assets/2152766/6689265/fc1ef946-ccb0-11e4-8551-59bc83a352e4.png)

In short, the GUI receives a full representation of what plug-ins are available, and what instances were found by running the discovered `Selector` plug-ins.

When the user then decides to initiate a publish, the GUI send a `advance` signal to the host saying..

> Hey Host, start processing the next plug-in in your list of plug-ins.

Once done, the results of the process is returned to the GUI for visualisation.

### The problem

..with this is that the GUI cannot visualise a process until *after* it has completed. The GUI is merely saying "host, process the **next** plug-in" without being too specific about which plug-in it means and in return it will get "gui, here's what I processed".

This is only a visual problem, as of yet. The real hassle involves our publishing logic. That is, when all validators have run and one of them produced errors, we would like processing to *not* continue to extractors. This is the `check()` from the diagram above. It looks something like this.

```python
def check():
  return not has_failures and current_plugin.type != “Extractor”
```

But, as we have already established, it should really look like this.

```python
def check():
  return not has_failures and next_plugin.type != “Extractor”
```

It needs to know about which plug-in is *about* to process, in order to make an informed decision about whether or not to send the next `advance()` signal.

### Possible Solution 1

We could compute the next plug-in from the host, by looking up at which index we are in the list of plug-ins, and simply look ahead to the next index.

```yaml
0: pluginA <-- "current"
1: pluginB 
2: pluginC
```

Once the next index is unavailable, we know that we have reached the end and that there aren't any other plug-ins to process.

However, there is a major downside to this approach; which is that we must, yet can't be, absolutely certain that the order in which plug-ins are represented in the list of plug-ins in the host is identical to the one in our GUI.

The reason this is difficult is because some plug-ins aren't visualised in the GUI - like Selectors, or those without an incompatible instance. Not to mention when and if we ever want to allow users to re-arrange items prior to publishing; so as to gain better control over the order of execution.

### Possible Solution 2

We could give the GUI absolute control over order of execution by passing a pair of items (plugin, instance) at a time.

![image](https://cloud.githubusercontent.com/assets/2152766/6689389/bee676f2-ccb1-11e4-8a5a-a09a09efcc0e.png)

This is the behaviour prior to State and have had it's own share of issues; mainly related to misuse of this power.

1. We could send a combination of (plugin, instance) that was *not* compatible; resulting in falsely processed items, visually.
2. We needed to rely on strict visual ordering, as any sorting or exclusion affects final processing.

Both of these can be solved by taking better care of the responsibility and may be what we need to do.

### Interface

Though no longer RESTful, this is a proposed interface.

```yaml
POST /state: Create state; discover plug-ins and run `Selectors`
GET /state: Compute and retrieve current state
POST /state <data>: Send changes
PUT /state <data>: Send (plugin, instance) pair to process.
```

**Example**

```python
request("POST", "/state")
state = request("GET", "/state")
# Draw
# User interacts
# Publish

errors = list()
for plugin in plugins:
  if plugin.flags & plugin.Disabled:
    continue

  if errors and type(plugin) == Extractor:
     break

  for instance in instances:
    if instance.changes:
      request("POST", "/state", data=changes)

    result = request("PUT", "/state", data={"plugin": plugin, "instance": instance})
    # blocks..
    if result.get("error"):
       errors.append(result["error"])    
```

### History

Just so that we have a sense of where we're coming from, here's how things have led to this point so far.

The first event-loop of Pyblish was text-based and written in Python. It processed each plug-in and each instance and dumped output to the console. Due to the similarities between the event-loop and how instances were processed within a single plug-in, responsibilities were mixed. What should have been purely event-loop decisions also became decisions made internal to the Pyblish processing mechanism. Something that can [still be seen today][legacy1].

In our first attempt at a GUI, the event-loop was programmed in JavaScript and hosted in QML. Shortly thereafter, I killed that approach due to the inherent complexity of maintaining this loop in JavaScript (mainly due to my inexperience) but more importantly due to the fact that I were mixing business logic with the presentation layer.

We have since then separated logic out of QML and into Python. Logic was restored and much of what was remained the same; the GUI merely hosted string-based names of plug-ins and instances and sent them across to the host one by one for processing. What I failed to realise at the time was that requests are blocking and that that could be used to determine when processing had finished in the host. Instead, a request was sent - the response simply saying "yay" or "nay" - and a polling mechanism was initiated; querying the host at even intervals for completion.

```yaml
Are you done yet?
Are you done yet?
Are you done yet?
...
```

This polling cause a tremendous performance bottleneck, as some plug-ins finished instantly (<2ms) and yet still had to wait for the polling to reach it's next iteration (~200ms).

In response to this, I delegated the responsibility to the host. I allowed the host to process as fast as it could and simply fed whatever information it spit out to the GUI as it went.

This is where we are today - before State - and things function well. However the GUI lost some of its influence of the processing which is a pity because this is what the GUI's was designed for to begin with.

Then State came to mind.

> What if we could fully represent State, and send back a modified copy, prior to letting the host crunch away as fast as it could?

Sounded pretty good to me, but yet again, here we are at a lack of control and obscure bugs popping up all over the place.

By now, I think what I need to do is step back and re-implement the scenario in which the GUI sent specific information about what to process - in the form of a (plugin, instance) pair and use the fact that requests block in order to send the next pair as fast as possible.

As far as I can tell, the results be equally fast, yet fully controlled.

[legacy1]: https://github.com/pyblish/pyblish/blob/c403257fb519f90557860a070c706ba5707342ed/pyblish/plugin.py#L250

**Comment by mottosso** [comment]
### Tracking Changes 3

With the above model of sending a pair of `(plugin, instance)`, when do we transmit and apply changes?

We could either send changes along with each pair; meaning updates are applied incrementally as plug-ins are processed. However this may leave the host in an undefined state, in which some changes from the GUI may have been applied but not others, depending on which plug-ins were actually processed. The behaviour would not be *atomic*.

Instead, we could apply changes much like we are currently, and *then* send *only* a pair of `(plugin, instance)` across. More or less ending up at where we left off before commencing this issue, with the addition of maintaining and applying changes.

![image](https://cloud.githubusercontent.com/assets/2152766/6704762/20a3dd2e-cd45-11e4-98ba-8d954799da70.png)

In this diagram, changes are applied separately from publishing. This leaves room for interactively saving changes, or to introduce an auto-save mechanism to trigger at a certain interval.

Commencing a publish with unsaved changes would simply transmit the changes first, await it's message of success, followed by the actual request to publish.

This also means that logic (the `check()` from above) is now fully encapsulated in the gui, the host merely doing as it is told, no questions asked.

**Comment by mottosso** [comment]
Implemented in 0.2.6
